include/exec/system_context.hpp

```diff
@@ -0,0 +1,433 @@
+      }
+
+      static void __cb(void* __data, int __completion_type, void* __exception) {
+        auto __self = static_cast<__op*>(__data);
+        __detail::__pass_to_receiver(__completion_type, __exception, std::move(__self->__recv_));
+      }
+
+      /// Object that receives completion from the work described by the sender.
+      __R __recv_;
+      /// The underlying implementation of the scheduler.
+      __exec_system_scheduler_interface* __scheduler_{nullptr};
+    };
+
+    /// Connects `__self` to `__r`, returning the operation state containing the work to be done.
+    template <class __R>
+    friend auto tag_invoke(stdexec::connect_t, system_sender&& __self, __R&& __r) noexcept(
+      std::is_nothrow_constructible_v<std::remove_cvref_t<__R>, __R>)
+      -> __op<system_sender, std::remove_cvref_t<__R>> {
+
+      return {std::move(__r), __self.__scheduler_};
+    }
+
+    /// Gets the environment of this sender.
+    friend __system_scheduler_env
+      tag_invoke(stdexec::get_env_t, const system_sender& __self) noexcept {
+      return {__self.__scheduler_};
+    }
+
+    /// The underlying implementation of the system scheduler.
+    __exec_system_scheduler_interface* __scheduler_{nullptr};
+  };
+
+  /// The state needed to execute the bulk sender created from system context.
+  template <stdexec::sender __Previous, std::integral __Size, class __Fn, class __R>
+  struct __bulk_state {
+    /// The sender object that describes the work to be done.
+    system_bulk_sender<__Previous, __Size, __Fn> __snd_;
+    /// The receiver object that receives completion from the work described by the sender.
+    __R __recv_;
+    /// The arguments passed from the previous receiver to the function object of the bulk sender (type-erased).
+    char* __arguments_data_{nullptr};
+  };
+
+  /// Receiver that is used in "bulk" to connect toe the input sender of the bulk operation.
+  template <stdexec::sender __Previous, std::integral __Size, class __Fn, class __R>
+  struct __bulk_intermediate_receiver {
+    /// Declare that this is a `receiver`.
+    using receiver_concept = stdexec::receiver_t;
+
+    /// The type of the object that holds relevant data for the entire bulk operation.
+    using __bulk_state_t = __bulk_state<__Previous, __Size, __Fn, __R>;
+
+    /// Object that holds the relevant data for the entire bulk operation.
+    __bulk_state_t& __state_;
+
+    /// Invoked when the previous sender completes with a value to trigger multiple operations on the system scheduler.
+    template <class... __As>
+    friend void tag_invoke(
+      stdexec::set_value_t,
+      __bulk_intermediate_receiver&& __self,
+      __As&&... __as) noexcept {
+      // Heap allocate input data in shared state as needed
+      auto __size = sizeof(std::tuple<__As...>);
+      __self.__state_.__arguments_data_ = new char[__size];
+      new (__self.__state_.__arguments_data_) std::tuple<__As...>{__as...};
```

Comment:
why are we heap-allocating this? why can't it be part of the operation state?

Reply:
Also, we should be storing the decayed values in the tuple.

Reply:
We shouldn't allocate here; we should be able to tell size of all the possible `set_value` arguments when we build this object. This is on my list of TODOs, but didn't get to it yet.

On the decayed values, I didn't noticed that. Thank you for pointing it out.